libbeat/common/datetime.go

```diff
@@ -21,37 +21,97 @@ import (
 	"encoding/binary"
 	"encoding/json"
 	"errors"
+	"fmt"
 	"hash"
 	"time"
 
 	"github.com/elastic/beats/v7/libbeat/common/dtfmt"
+	conf "github.com/elastic/elastic-agent-libs/config"
 )
 
 const (
+	millisecPrecision TimestampPrecision = iota + 1
+	microsecPrecision
+	nanosecPrecision
+
+	DefaultTimestampPrecision = millisecPrecision
+
 	// TsLayout is the seconds layout to be used in the timestamp marshaling/unmarshaling everywhere.
 	// The timezone must always be UTC.
 	TsLayout = "2006-01-02T15:04:05.000Z"
 
 	tsLayoutMillis = "2006-01-02T15:04:05.000Z"
 	tsLayoutMicros = "2006-01-02T15:04:05.000000Z"
 	tsLayoutNanos  = "2006-01-02T15:04:05.000000000Z"
+
+	millisecPrecisionFmt = "yyyy-MM-dd'T'HH:mm:ss.fff'Z'"
+	microsecPrecisionFmt = "yyyy-MM-dd'T'HH:mm:ss.ffffff'Z'"
+	nanosecPrecisionFmt  = "yyyy-MM-dd'T'HH:mm:ss.fffffffff'Z'"
+)
+
+var (
+	defaultParseFormats = []string{
+		tsLayoutMillis,
+		tsLayoutMicros,
+		tsLayoutNanos,
+	}
+
+	precisions = map[TimestampPrecision]string{
+		millisecPrecision: millisecPrecisionFmt,
+		microsecPrecision: microsecPrecisionFmt,
+		nanosecPrecision:  nanosecPrecisionFmt,
+	}
+
+	timeFormatter = dtfmt.MustNewFormatter(precisions[DefaultTimestampPrecision])
 )
 
 // Time is an abstraction for the time.Time type
 type Time time.Time
 
-var defaultTimeFormatter = dtfmt.MustNewFormatter("yyyy-MM-dd'T'HH:mm:ss.fffffffff'Z'")
+type TimestampPrecision uint8
+
+type TimestampConfig struct {
+	Precision TimestampPrecision `config:"precision"`
+}
+
+func defaultTimestampConfig() TimestampConfig {
+	return TimestampConfig{Precision: DefaultTimestampPrecision}
+}
+
+func (p *TimestampPrecision) Unpack(v string) error {
+	switch v {
+	case "millisecond", "":
+		*p = millisecPrecision
+	case "microsecond":
+		*p = microsecPrecision
+	case "nanosecond":
+		*p = nanosecPrecision
+	default:
+		return fmt.Errorf("invalid timestamp precision %s, available options: millisecond, microsecond, nanosecond", v)
+	}
+	return nil
+}
+
+func SetTimestampPrecision(c *conf.C) error {
```

Comment:
This seems important and I think it needs some documentation. Like it should state that it modifies the behavior of `common.Time`'s `MarshalJSON()` and `String()` methods . And I think it should call out that it's not safe to call this after you have any other threads using `common.Time` because it modifies the global formatter that they all share.